Optimize AD residual and Jacobian object contributions

GiudGiud opened the pull request:
seems this gains the 5% for February. This only improves AD computations though, and likely not FV

GiudGiud committed:
Remove assert loop and if condition from opt mode for sure
refs #10256

GiudGiud committed:
Minimize AD computations in AD objects
refs #10256

GiudGiud committed:
Avoid residual vector allocations for ADKernels
refs #10256

GiudGiud commented:
Without changes. 
```


Currently Executing
  Performing Initial Setup
    Projecting Initial Solutions                                                         [  8.59 s] [ 8251 MB]
    Finished Updating ActiveSemiLocalNode Range                                          [  0.85 s] [ 8460 MB]
    Currently Building SemiLocalElemMap                                                  [  8.06 s] [ 8735 MB]
  Finished Performing Initial Setup                                                      [ 20.45 s] [ 9035 MB]
  Currently Outputting
    Outputting Step                                                                      [  5.10 s] [ 9854 MB]
  Finished Outputting                                                                    [  5.10 s] [ 9854 MB]

    Computing Residual                                                                   [  6.64 s] [10192 MB]
 0 Nonlinear |R| = 6.399219e+01
    Computing Jacobian....                                                               [ 28.30 s] [11804 MB]
      0 Linear |R| = 6.399219e+01
      1 Linear |R| = 2.817763e+00
      2 Linear |R| = 1.111162e-01
      3 Linear |R| = 5.700842e-03
      4 Linear |R| = 2.657560e-04

    Computing Residual                                                                   [  6.96 s] [15429 MB]
 1 Nonlinear |R| = 2.657560e-04
    Computing Jacobian....                                                               [ 28.78 s] [15411 MB]
      0 Linear |R| = 2.657560e-04
      1 Linear |R| = 6.330617e-05
      2 Linear |R| = 3.604962e-06
      3 Linear |R| = 2.026304e-07
      4 Linear |R| = 2.071699e-08
      5 Linear |R| = 6.766795e-10

    Computing Residual                                                                   [  6.68 s] [14518 MB]
 2 Nonlinear |R| = 6.766796e-10
 Solve Converged!
  Finished Solving                                                                       [107.07 s] [ 9045 MB]
  Currently Outputting
    Finished Outputting Step                                                             [  1.98 s] [ 9438 MB]
    Finished Outputting Step                                                             [  2.96 s] [ 9573 MB]
  Finished Outputting                                                                    [  4.94 s] [ 9573 MB]
Finished Executing                                                                       [137.57 s] [ 9573 MB]
```


with changes

```
(moose) giudgl@fn434257:~/projects/moose_v1/test/tests/kernels/ad_value$ ../../../moose_test-opt -i ad_value.i -r 10
Setting Up
  Uniformly Refining
    Finished Updating Mesh                                                               [  3.39 s] [ 2548 MB]
  Finished Uniformly Refining                                                            [ 12.98 s] [ 2548 MB]
  Initializing
    Finished Updating Because Mesh Changed                                               [  3.64 s] [ 2947 MB]
    Currently Initializing Equation Systems..                                            [ 16.05 s] [ 7912 MB]
  Finished Initializing                                                                  [ 19.69 s] [ 7912 MB]
Finished Setting Up                                                                      [ 33.23 s] [ 7912 MB]
Framework Information:
MOOSE Version:           git commit d540445fca on 2025-03-03
LibMesh Version:         
PETSc Version:           3.22.1
SLEPc Version:           3.22.1
Current Time:            Mon Mar  3 09:56:15 2025
Executable Timestamp:    Mon Mar  3 09:54:28 2025

Input File(s):
  /Users/giudgl/projects/moose_v1/test/tests/kernels/ad_value/ad_value.i

Command Line Argument(s):
  -r=10

Checkpoint:
  Wall Time Interval:      Every 3600 s
  User Checkpoint:         Disabled
  # Checkpoints Kept:      2
  Execute On:              TIMESTEP_END 

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   4198401
  Elems:                   4194304
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                8396802
  Num Local DOFs:          8396802
  Variables:               { "u" "v" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                8396802
  Num Local DOFs:          8396802
  Variables:               { "u_jac" "v_jac" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             NEWTON
  PETSc Preconditioner:    hypre boomeramg 
  MOOSE Preconditioner:    SMP (auto)

Currently Executing
  Performing Initial Setup
    Finished Projecting Initial Solutions                                                [  8.42 s] [ 7913 MB]
    Finished Updating ActiveSemiLocalNode Range                                          [  0.72 s] [ 8115 MB]
    Currently Building SemiLocalElemMap                                                  [  7.53 s] [ 8354 MB]
  Finished Performing Initial Setup                                                      [ 19.42 s] [ 8625 MB]
  Currently Outputting
    Finished Outputting Step                                                             [  4.75 s] [ 9436 MB]
  Finished Outputting                                                                    [  4.75 s] [ 9436 MB]

    Computing Residual                                                                   [  6.31 s] [ 9774 MB]
 0 Nonlinear |R| = 6.399219e+01
    Computing Jacobian...                                                                [ 26.82 s] [10393 MB]
      0 Linear |R| = 6.399219e+01
      1 Linear |R| = 2.817763e+00
      2 Linear |R| = 1.111162e-01
      3 Linear |R| = 5.700842e-03
      4 Linear |R| = 2.657560e-04

    Computing Residual                                                                   [  6.65 s] [15341 MB]
 1 Nonlinear |R| = 2.657560e-04
    Computing Jacobian....                                                               [ 27.38 s] [15120 MB]
      0 Linear |R| = 2.657560e-04
      1 Linear |R| = 6.330617e-05
      2 Linear |R| = 3.604962e-06
      3 Linear |R| = 2.026304e-07
      4 Linear |R| = 2.071699e-08
      5 Linear |R| = 6.766795e-10
    Computing Residual                                                                   [  6.33 s] [14514 MB]
 2 Nonlinear |R| = 6.766796e-10
 Solve Converged!
  Finished Solving                                                                       [101.63 s] [ 8961 MB]
  Currently Outputting
    Finished Outputting Step                                                             [  1.79 s] [ 9357 MB]
    Finished Outputting Step                                                             [  2.80 s] [ 9492 MB]
  Finished Outputting                                                                    [  4.59 s] [ 9492 MB]
Finished Executing                                                                       [130.41 s] [ 9492 MB]

```
why am I not using Instruments? Ask 6-1000

lindsayad reviewed (approved):
The Jacobian failure in solid mechanics looks like a red herring. All the ratio tolerances are less than 2e-8

GiudGiud committed:
Avoid cast errors with displaced meshes
and postprocessors
Slightly raise tolerance on jac test, which is still very tight and error is likely from floating point precision

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2725820/), step Docs: sync website on 93bc2e7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30004/site)

This comment will be updated on new commits.

dschwen commented on framework/src/bcs/ADIntegratedBC.C:
```diff
@@ -99,14 +99,12 @@ ADIntegratedBCTempl<T>::computeResidual()
   for (auto & r : _residuals)
     r = 0;
 
-  if (_use_displaced_mesh)
-    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-      for (_i = 0; _i < _test.size(); _i++)
-        _residuals[_i] += raw_value(_ad_JxW[_qp] * _ad_coord[_qp] * computeQpResidual());
-  else
-    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-      for (_i = 0; _i < _test.size(); _i++)
-        _residuals[_i] += raw_value(_JxW[_qp] * _coord[_qp] * computeQpResidual());
```
This is funny. First we specifically distinguish between needing AD or not and then we raw_value both cases anyways...

lindsayad replied:
I'm glad my silly code gives you giggles. Probably at some point I considered that only the `_ad_JxW` reference might be valid or properly computed in `Assembly` when using AD on the displaced mesh

lindsayad replied:
Or perhaps this is a remnant of the days when AD kernels were templated on the `ComputeType`

GiudGiud committed:
Slightly raise tolerances on 3 more solid mechanics tests in parallel
same order of magnitude tight, and they were close to their tolerance already so FP computation errors can explain why they now fail
